# MATH 468 - Stochastic Calculus II - HW 3

In [1]:
'''Packages Used'''
import pandas as pd
import numpy as np

## Problem 1

We consider a two-period binomial model with the following properties: each period lasts a year and the current stock price is $S_0 = 4$. On each period, the stock price doubles when it moves up and is reduced by half when it moves down. The annual interest rate on the money market is $25\%$. (This model is the same as in Prob. 1 of HW#2).  
We consider four options on this market:  
* a *European Call* option with maturity $T=2$ years and strike price $K=4$.  
* a *European Put* option with maturity $T=2$ years and strike price $K=4$.  
* an *American Call* option with maturity $T=2$ years and strike price $K=4$.  
* an *American Put* option with maturity $T=2$ years and strike price $K=4$.  

(a) Find the price at time $0$ of both European options.  
(b) Find the price at time $0$ of both American options. Compare your results with (a) and comment.  
(c) For each of the american options, describe the optimal excerising strategy.  
(d) We assume that you sell the American put to a market participant A for the price found in $b$.  Explain how you act on the market to be able to payoff A when he excercises the option.  (We assume that A exercises optimally)  
(e) Now assume that A makes a mistake and doesn't excercise optimally.  What is the consequence for you as the seller of the option? (Only a short answer is required)

### Solution to Part (a)

To derive the price of the option, we begin by understanding the option payoff at expiry.

| $$i$$ | $$S_i$$ | $$f_i^c$$ | $$f_i^p$$ |
| -- | -- | -- | -- |
| $$7$$ | $$16$$ | $$12$$ | $$0$$ |
| $$5\&6$$ | $$4$$ | $$0$$ | $$0$$ |
| $$4$$ | $$1$$ | $$0$$ | $$3$$ |

Using these payoffs and the risk-neutral probability $q=\frac{1}{2}$ derived in HW 2, we calculate the prices of the european call and put below.

**European Call**  
$$f_3^c = \frac{1}{1 + 25\%}(\frac{1}{2}(12) + \frac{1}{2}(0)) = \frac{4}{5}(6) = \frac{24}{5}$$
$$f_2^c = \frac{1}{1 + 25\%}(\frac{1}{2}(0) + \frac{1}{2}(0)) = 0$$
$$f_1^c = \frac{1}{1 + 25\%}(\frac{1}{2}(\frac{24}{5}) + \frac{1}{2}(0)) = \frac{4}{5}(\frac{12}{5}) = \frac{48}{25}$$

**European Put**  
$$f_3^p = \frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(0)) = 0$$
$$f_2^p = \frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(3)) = \frac{4}{5}(\frac{3}{2}) = \frac{6}{5}$$
$$f_2^p = \frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(\frac{6}{5})) = \frac{4}{5}(\frac{3}{5}) = \frac{12}{25}$$

### Solution to Part (b)

We adjust the calculations to reflect the nature of the American Option's early expiry

**American Call**  
$$f_3^C = \max\{\frac{1}{1 + 25\%}(\frac{1}{2}(12) + \frac{1}{2}(0)),8-4\} = \max\{\frac{4}{5}(6),4\} = \frac{24}{5}$$
$$f_2^C = \max\{\frac{1}{1 + 25\%}(\frac{1}{2}(0) + \frac{1}{2}(0)),0\} = 0$$
$$f_1^C = \max\{\frac{1}{1 + 25\%}(\frac{1}{2}(\frac{24}{5}) + \frac{1}{2}(0)),0\} = \frac{4}{5}(\frac{12}{5}) = \frac{48}{25}$$

**American Put**  
$$f_3^P = \max\{\frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(0)),0\} = 0$$
$$f_2^P = \max\{\frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(3)),2\} = \max\{\frac{4}{5}(\frac{3}{2}),2\} = 2$$
$$f_2^P = \max\{\frac{1}{1+25\%}(\frac{1}{2}(0) + \frac{1}{2}(2)),0\} = \frac{4}{5}$$

### Solution to Part (c)

For the american call, it is optimally excercised at expiry, as shown in the calculations.  For the american put, it is optimally excercised at time $t=1$.  This is because the down move price relative to the strike price difference is greater than the present value of the risk-neutral expected value.  To summarize, the call should be held to expiry while the put should be excercised early.

### Solution to Part (d)

Market Participant A will only exercise the option at time $t=1$ after a down move in the stock.  If the stock moves up, the option becomes worthless.  I included a table of cashflows below.

| Time | $$t=0$$ | Down Move | Up Move |
| -- | -- | -- | -- |
| Short Put | $$\frac{4}{5}$$ | $$-2$$ | $$0$$ |
| Short Delta | $$4$$ | $$-2$$ | $$-8$$
| Long Rates | $$-\frac{24}{5}$$ | $$6$$ | $$6$$ |
| **Cash Flows** | $$0$$ | $$2$$ | $$-2$$ |

This combination makes us effectively short the stock but with downside cashflows hedged to only lose 2 dollars instead of 4.

### Solution to Part (e)

In the event the stock initially rises to 8, the put becomes worthless and we close out the trade with the same $-2$ cashflow shown in Part (d).  We see that no exercise with the initial down move will lead to the following cashflows, which is greater than the previously derived cashflow of $2$.  To summarize, this works better for us.

| Time | $$t=0$$ | $S_2=1$ | $S_2=4$ |
| -- | -- | -- | -- |
| Short Put | $$\frac{4}{5}$$ | $$-3$$ | $$0$$ |
| Short Delta | $$4$$ | $$-1$$ | $$-4$$ |
| Long Rates | $$-\frac{24}{5}$$ | $$\frac{15}{2}$$ | $$\frac{15}{2}$$ |
| **Cash Flows** | $$0$$ | $$\frac{7}{2}$$ | $$\frac{7}{2}$$ |


## Problem 2 (Problem 1 continued)

We consider the same stock evolution as in Problem 1, but now assume the market is 3-period and the expiration date of the options is $T=3$ years.  
(a) Repeat questions (a), (b), and (c) of Problem 1 in this new market.  (You don't have to provide all the calculations, numerical results are sufficient)  
(b) We consider a new type of option on this market: an *American straddle* option with strike price $K=4$. The payoff of an American straddle is the sum of the payoffs of an American call and an American put. If you exercise it when $S_t=5$, your payoff is $1$. If you exercise it when $S_t=2$, your payoff is $2$. You are allowed to exercise at any time. Find the price of this option.  
(c) Describe the optimal excercising strategy.  
(d) Let $C_0$ (resp. $P_0$, $K_0$) be the prices of the American call (resp. put, straddle).  Give an intuitive explanation why you found that
$$K_0 < C_0 + P_0$$

### Solution to Part (a)

As in Problem 1, we begin by understanding the payoff function of each option

| $$i$$ | $$S_i$$ | $$f_i^c$$ | $$f_i^p$$ |
| -- | -- | -- | -- |
| $$15$$ | $$32$$ | $$28$$ | $$0$$ |
| $$11,13,\&14$$ | $$8$$ | $$4$$ | $$0$$ |
| $$9,10,\&12$$ | $$2$$ | $$0$$ | $$2$$ |
| $$8$$ | $$\frac{1}{2}$$ | $$0$$ | $$\frac{7}{2}$$ |

Using the payoff function, we calculate the price of each option as follows  

**European Call**  
$$f_7^c = \frac{4}{5}(\frac{1}{2}(28) + \frac{1}{2}(4)) = \frac{64}{5}$$
$$f_{5\&6}^c = \frac{4}{5}(\frac{1}{2}(4) + \frac{1}{2}(0)) = \frac{8}{5}$$
$$f_4^c = \frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(0)) = 0$$
$$f_3^c = \frac{4}{5}(\frac{1}{2}(\frac{64}{5}) + \frac{1}{2}(\frac{8}{5})) = \frac{144}{25}$$
$$f_2^c = \frac{4}{5}(\frac{1}{2}(\frac{8}{5}) + \frac{1}{2}(0)) = \frac{16}{25}$$
$$f_1^c = \frac{4}{5}(\frac{1}{2}(\frac{144}{25}) + \frac{1}{2}(\frac{16}{25})) = \frac{64}{25}$$

**European Put**  
$$f_7^p = \frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(0)) = 0$$
$$f_{5\&6}^p = \frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(2)) = \frac{4}{5}$$
$$f_4^p = \frac{4}{5}(\frac{1}{2}(2) + \frac{1}{2}(\frac{7}{2})) = \frac{11}{5}$$
$$f_3^p = \frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(\frac{4}{5})) = \frac{8}{25}$$
$$f_2^p = \frac{4}{5}(\frac{1}{2}(\frac{4}{5}) + \frac{1}{2}(\frac{11}{5})) = \frac{6}{5}$$
$$f_1^p = \frac{4}{5}(\frac{1}{2}(\frac{8}{25}) + \frac{1}{2}(\frac{6}{5})) = \frac{76}{125}$$

**American Call**  
By the theorem of american option function convexity, we see the price of the american call is the same as the european.  Therefore
$$f_1^C = f_1^c = \frac{64}{25}$$
I also validated this result but choose to leave it out for the sake of brevity

**American Put**  
$$f_7^P = \max\{\frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(0)),0\} = 0$$
$$f_{5\&6}^P = \max\{\frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(2)),0\} = \max\{\frac{4}{5},0\} = \frac{4}{5}$$
$$f_4^P = \max\{\frac{4}{5}(\frac{1}{2}(2) + \frac{1}{2}(\frac{7}{2})),3\} = \max\{\frac{11}{5},3\} = 3$$
$$f_3^P = \max\{\frac{4}{5}(\frac{1}{2}(0) + \frac{1}{2}(\frac{4}{5})),0\} = \max\{\frac{8}{25},0\} = \frac{8}{25}$$
$$f_2^P = \max\{\frac{4}{5}(\frac{1}{2}(\frac{4}{5}) + \frac{1}{2}(3)),2\} = \max\{\frac{76}{50},2\} = 2$$
$$f_1^P = \frac{4}{5}(\frac{1}{2}(\frac{8}{25}) + \frac{1}{2}(2)) = \frac{116}{125}$$

Clearly, it is best to exercise the american put early after an initial down move.  If the inital move is up, then it is best to wait to hope for a move back down and back down again.  This is because an up then down move brings the stock price to $4$, which makes the option worthless still.

### Solution to Part (b)

We begin to solve for the value of this option by understanding the payoff function, which is listed in the table below.

| $$i$$ | $$S_i$$ | $$f_i^C$$ | $$f_i^P$$ | $$f_i^K$$ |
| -- | -- | -- | -- | -- |
| $$15$$ | $$32$$ | $$28$$ | $$0$$ |$$28$$ |
| $$11,13,\&14$$ | $$8$$ | $$4$$ | $$0$$ | $$4$$ |
| $$9,10,\&12$$ | $$2$$ | $$0$$ | $$2$$ | $$2$$
| $$8$$ | $$\frac{1}{2}$$ | $$0$$ | $$\frac{7}{2}$$ | $$\frac{7}{2}$$ |

We also note that the straddle is american in nature, so we must also include max functions in calculating its value.  For these max functions, the "early exercise" value is $|S-K|$, the absolute value of the difference between the stock price and the strike price

**American Straddle**
$$f_7^K = \max\{\frac{4}{5}(\frac{1}{2}(28) + \frac{1}{2}(4)),|16 - 4|\} = \max\{\frac{64}{5},12\} = \frac{64}{5}$$
$$f_{5\&6}^K = \max\{\frac{4}{5}(\frac{1}{2}(4) + \frac{1}{2}(2)),|4-4|\} = \max\{\frac{12}{5},0\} = \frac{12}{5}$$
$$f_4^K = \max\{\frac{4}{5}(\frac{1}{2}(2) + \frac{1}{2}(\frac{7}{2})),|1-4|\} = \max\{\frac{11}{5},3\} = 3$$
$$f_3^K = \max\{\frac{4}{5}(\frac{1}{2}(\frac{64}{5}) + \frac{1}{2}(\frac{12}{5})),|8-4|\} = \max\{\frac{304}{50},4\} = \frac{152}{25}$$
$$f_2^K = \max\{\frac{4}{5}(\frac{1}{2}(\frac{12}{5}) + \frac{1}{2}(3)),|2-4|\} = \max\{\frac{54}{25},2\} = \frac{54}{25}$$
$$f_1^K = \frac{4}{5}(\frac{1}{2}(\frac{152}{25}) + \frac{1}{2}(\frac{54}{25})) = \frac{412}{125}$$

### Solution to Part (c)

The optimal exercise strategy is to hold the straddle until expiry unless the stock price cannot recover to the strike.  This is easily seen in the pricing arithmetic as the max function only selects the difference between strike and stock price from node 4, the bottom most node at time $t=2$.  Put more simply, hold the straddle so long as their is a chance that the call-end of it will be in-the-money at expiry, else excercise the put.

### Solution to Part (d)

We have calulated $C_0=\frac{64}{25}$, $P_0=\frac{116}{125}$, and $K_0=\frac{412}{125}$.  We can easily see
$$C_0 + P_0 = (\frac{5}{5})(\frac{64}{25}) + \frac{116}{125} = \frac{436}{125} > \frac{412}{125} = K_0$$
So the given equation holds.  This is due to the fact that in the valuation of the call and put, the owner will exercise their option earlier in the event of a down move than if they hold the straddle, where the optimal strategy is to wait for 2 down moves before exercising.  The earlier potential value in the put makes the individual securities value summed greater than that of their combination.

## Problem 3 (*N*-period model with *N* large)

*Note*: The purpose of this problem below is to use computational techniques (Excel spreadsheet, Matlab, R, Python, etc.) and code the dynamic programming ideas seen in class. Please provide the numerical answer to the questions as well as a sample of your work (spreadsheet, code file, etc.).

We consider an *N*-period binomial model with the following properties: $N=60$, the current stock price is $S_0=1000$; on each period, the stock price increases by $0.5\%$ when it moves up and decreases by $0.3\%$ when it moves down. The annual interest rate on the money market is $5\%$. (Notice that this model is a CRR model, which means that the binomial tree is recombining.)

(a) Find the price at time $t_0=0$ of a (European) call option with strike price $K=1025$ and maturity $T=1$ year.  
(b) Find the price at time $t_0=0$ of a (European) put option with strike price $K=1025$ and maturity $T=1$ year.  
(c) We consider now, that you are at time $t_5$ (i.e. after 5 periods, which represents 1 month later). Assume that the stock went up for the first three periods and then down for the next two. Determine the price of the stock at that position, the price of each of the put and call and determine how many shares of stock you should be holding in the hedging portfolio for each of the call and the put.  
(d) Repeat questions (a)-(c) in the case where $r=0$. Comment on the results.  
*Hint*: take advantage of the fact that the model is recombining: moving up, then down
takes you to the same value as moving first down, and then up. It reduces the amount of
calculations necessary and will speed up your code.

*Note*: In the following sections, I write the functionality used to solve the parts listed.  I add comments to the code showing where I intend to use the functionality.

In [2]:
def walk_params(u,d,T,N):
    '''get time step'''
    dt = T/N

    '''get mean'''
    mu = (np.log(1+u) + np.log(1-d)) / (2*dt)
    
    '''get variance'''
    sigma = (np.log(1+u) - mu*dt) / (np.sqrt(dt))

    return mu, sigma

In [3]:
def build_tree(N, T, mu, sigma, s0, k, r, option_type):
    '''Establish time delta -- each variable represents
    T - time to maturity in years
    N - number of periods (i.e. time steps)'''
    dt = T / N

    '''Establish up and down factos'''
    u = np.exp(mu*dt + sigma*np.sqrt(dt))
    d = np.exp(mu*dt - sigma*np.sqrt(dt))
    
    '''Establish risk neutral probability'''
    q = ((1+r*dt) - d) / (u - d)
    
    '''Constructs the data frame -- each column and its representation
    index - number
    layer - time step
    stock_price - stock price at that node and layer
    phi - shares in replicating protfolio
    psi - amount of funds in money market
    option value - option price at that node'''
    tree = pd.DataFrame(data=None, index=range(0,sum(range(N+2))))
    tree.index.name = 'node'
    tree['layer'] = ""
    tree['stock_price'] = 0
    tree['phi'] = 0
    tree['psi'] = 0
    
    '''inputs intial stock price'''
    tree['stock_price'].loc[0] = s0

    '''Inputs layers (time steps) and further stock prices'''
    node = 0
    for i in range(0,N+2):
        if i < (N+1):
            tree.iloc[range(node,node+i+1),0] = i #Layer
            tree.iloc[range(node,node+i+1),1] = s0 * d ** (np.arange(i,-1,-1)) * u ** (np.arange(0,i+1,1)) #Stock Price
            node += i + 1

    '''Puts intrinsic option value in at every node'''
    if option_type == 'c':
        tree['option_value'] = tree['stock_price'].apply(lambda x: x-k if x-k > 0 else 0)
    elif option_type == 'p':
        tree['option_value'] = tree['stock_price'].apply(lambda x: k-x if k-x > 0 else 0)
    
    '''resets node to N-1'''
    
    node= node - N
    
    '''Calculates option value and replicating portfolio going backwards'''
    for i in reversed(range(N)):
        tree.iloc[range(node-i-2,node-1),4] = (1 / (1+r*dt)) * (q * tree.iloc[range(node,node+i+1),4].to_numpy() + (1-q) * tree.iloc[range(node-1,node+i),4].to_numpy()) # price option
        tree.iloc[range(node-i-2,node-1),2] = (tree.iloc[range(node,node+i+1),4].to_numpy() - tree.iloc[range(node-1,node+i),4].to_numpy()) / (tree.iloc[range(node,node+i+1),1].to_numpy() - tree.iloc[range(node-1,node+i),1].to_numpy()) # get delta
        tree.iloc[range(node-i-2,node-1),3] = (tree.iloc[range(node,node+i+1),4].to_numpy() - tree.iloc[range(node-i-2,node-1),2].to_numpy() * tree.iloc[range(node,node+i+1),1].to_numpy()) / (1 + r*dt) # Money Market
        node = node - i - 1


    return tree

### Solution to Part (a)

In [4]:
T=1 # 1 Year
N=60 # 60 Periods
mu, sigma = walk_params(u=0.005,d=0.003,T=T,N=N)
tree_c = build_tree(N=N, T=T, mu=mu, sigma=sigma, s0=1000, k=1025, r=.05, option_type='c')
tree_c.iloc[0,:]

C:\Users\jdean\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


layer                    0
stock_price         1000.0
phi               0.798915
psi            -770.378395
option_value     28.536695
Name: 0, dtype: object

The price of the option is $\$28.54$

### Solution to Part (b)

In [5]:
tree_p = build_tree(N=60, T=1, mu=mu, sigma=sigma, s0=1000, k=1025, r=.05, option_type='p')
tree_p.iloc[0,:]

layer                    0
stock_price         1000.0
phi              -0.201085
psi             204.652066
option_value      3.567157
Name: 0, dtype: object

The price of the option is $\$3.57$

### Solution to Part (c)

In [6]:
t5_tree_c = tree_c[tree_c['layer']==5]
t5_tree_p = tree_p[tree_p['layer']==5]

In [7]:
t5_tree_c

layer  stock_price       phi         psi  option_value
node                                                       
15       5   985.089730  0.588668 -565.012023     14.879245
16       5   992.994162  0.689458 -664.711662     19.915666
17       5  1000.962018  0.777898 -752.898292     25.748231
18       5  1008.993810  0.850065 -825.435329     32.275042
19       5  1017.090049  0.904785 -880.876844     39.370538
20       5  1025.251253  0.943301 -920.214314     46.905916

In [8]:
t5_tree_c.iloc[3,:]

layer                    5
stock_price     1008.99381
phi               0.850065
psi            -825.435329
option_value     32.275042
Name: 18, dtype: object

In [9]:
t5_tree_p.iloc[3,:]

layer                    5
stock_price     1008.99381
phi              -0.149935
psi             153.664536
option_value      2.381098
Name: 18, dtype: object

For this, time $t_5$ is denoted by layer 5, which I call in each of the previous code blocks.  We can see the stock is worth $\$1008.99$ with the call and put worth $\$32.28$ and $\$2.38$, respectively.  We can see the long call requires a hedged portfolio of $0.85$ shares and $\$825.43$ in the money market.  The long put requires $0.15$ shares and $\$153.66$ in the money market.  It is easily seen that the shares traded relative to the call and put adds to $1$, representing the greek delta of the option.

### Solution to Part (d)

We see the call is wortth $\$3.60$ and the put is worth $\$28.60$, a complete role reversal vs the $5\%$ interest in the previous case.  We see the absolute value of the $\phi$'s still sum to one in both time instances, indicating the put-call parity is holding.  The option statistics for time-step 5 are also included.

In [10]:
# repeat part a
r0_tree_c = build_tree(N=N, T=T, mu=mu, sigma=sigma, s0=1000, k=1025, r=0, option_type='c')
r0_tree_c.iloc[0,:]

C:\Users\jdean\AppData\Local\Programs\Python\Python39\lib\site-packages\pandas\core\indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


layer                  0
stock_price       1000.0
phi             0.217913
psi            -214.3112
option_value    3.602071
Name: 0, dtype: object

In [11]:
# repeat part b
r0_tree_p = build_tree(N=N, T=T, mu=mu, sigma=sigma, s0=1000, k=1025, r=0, option_type='p')
r0_tree_p.iloc[0,:]

layer                   0
stock_price        1000.0
phi             -0.782087
psi              810.6888
option_value    28.602071
Name: 0, dtype: object

In [12]:
#repeat part c
t5_r0_tree_c = r0_tree_c[r0_tree_c['layer']==6]
t5_r0_tree_p = r0_tree_p[r0_tree_p['layer']==6]
print('---------- call ----------')
print(t5_r0_tree_c.iloc[3,:])
print(' ')
print('---------- put ----------')
print(t5_r0_tree_p.iloc[3,:])

---------- call ----------
layer                     6
stock_price     1005.966828
phi                 0.26881
psi              -265.88962
option_value       4.524259
Name: 24, dtype: object
 
---------- put ----------
layer                     6
stock_price     1005.966828
phi                -0.73119
psi               759.11038
option_value       23.55743
Name: 24, dtype: object
